### PR TITLE
Hetero plugin supports ngraph

### PR DESCRIPTION
**Pull request by apankratovantonp**
* If `ngraph` function was passed as CNN network implementation hetero plugin preserve `ngraph` representation and pass sub graphs as `ngraph` functions to 
* New code dose not depend on `ADE`
* `HETERO` and `MULTI` dose not downgrade `cnnnetwork`
* `MKLDNN` accepts `ngraph` based `cnnnetwork` in query API

### CODE REVIEW

File: inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp [hunk 1]
```diff
@@ -203,18 +206,48 @@ void Engine::AddExtension(InferenceEngine::IExtensionPtr extension) {
...
-            std::unique_ptr <MKLDNNNode>(MKLDNNNode::CreateNode(*i, eng, extensionManager, fake_w_cache));
-            res.supportedLayersMap.insert({ (*i)->name, GetName() });
-        } catch (InferenceEngine::details::InferenceEngineException&) {
+    MKLDNNWeightsSharing::Ptr fake_w_cache;
+    auto function = network.getFunction();
+    if (function != nullptr) {
+        std::unordered_set<std::string> originalOps;
+        {
+            auto clonedNetwork = cloneNetwork(network);
+            for (details::CNNNetworkIterator itLayer{clonedNetwork.get()}; itLayer != details::CNNNetworkIterator(); itLayer++) {
```
Comment (ilya-lavrenov) [inline inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp:215]: `details::CNNNetworkIterator` leads to conversion to old representation.. `itLayer->name` is `CNNLayer`, while you should work with `ngraph::Op`

File: inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp [hunk 2]
```diff
@@ -203,18 +206,48 @@ void Engine::AddExtension(InferenceEngine::IExtensionPtr extension) {
...
+        }
+        auto clonedNetwork = cloneNetwork(network);
+        clonedNetwork = Transformation(clonedNetwork->getFunction(), clonedNetwork);
+        for (details::CNNNetworkIterator itLayer{clonedNetwork.get()}; itLayer != details::CNNNetworkIterator(); itLayer++) {
+            try {
+                mkldnn::engine eng(mkldnn::engine(mkldnn::engine::kind::cpu, 0));
+                // if we can create and have not thrown exception, then layer is supported
+                std::unique_ptr <MKLDNNNode>(MKLDNNNode::CreateNode(*itLayer, eng, extensionManager, fake_w_cache));
+                std::cout << (*itLayer)->name << ": ";
+                for (auto&& fusedLayerName : ngraph::getFusedNamesVector((*itLayer)->getNode())) {
```
Comment (ilya-lavrenov) [inline inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp:229]: @ilyachur  is it safe? Looks like we convert from opset3 to opset_legacy first. And after translate 1:1 to CNNLayers and getNode should work, is not it?

File: inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp (hunk 1 above)
Comment (apankratovantonp) [reply inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp:215]: Used just to get names from **cloned** network.

File: inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp (hunk 2 above)
Comment (apankratovantonp) [reply inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp:229]: Yes it works

Comment (ilyachur) [reply inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp:229]: Yes it should work. Conversions don't change the operation names.

**Comment by apankratovantonp** [comment]
@ilyachur Dose conversion to legacy representation applies any transformations and can redeuce opset of Network?

**Comment by ilyachur** [comment]
> Dose conversion to `CNNLayer` representation applies any transformations and can reduce opset of Network?

Yes, before the conversion we need to transform nGraph to legacy, when we do it, we reduce opset version to 1.

File: inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp (hunk 1 above)
Comment (apankratovantonp) [reply inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp]: removed

File: inference-engine/ie_bridges/python/tests/test_IECore.py [hunk 3]
```diff
@@ -49,6 +49,7 @@ def test_load_network_wrong_device():
     assert 'Device with "BLA" name is not registered in the InferenceEngine' in str(e.value)
 
 
+@pytest.mark.skip(reason="IENetwork.layers return not all layers in case of ngraph representation due to inner conversion into legacy representation")
```
Comment (ilyachur) [inline inference-engine/ie_bridges/python/tests/test_IECore.py:52]: Why did you skip these tests for all plugins?
When will it be enabled?

Comment (apankratovantonp) [reply inference-engine/ie_bridges/python/tests/test_IECore.py:52]: Python API Converts to old representation internally. There should be separate tests for legacy and ngraph representation or there should be no conversion in API.

Comment (ilyachur) [reply inference-engine/ie_bridges/python/tests/test_IECore.py:52]: Can you create a ticket for this activity?

File: inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp [hunk 4]
```diff
@@ -203,18 +206,42 @@ void Engine::AddExtension(InferenceEngine::IExtensionPtr extension) {
...
-            // if we can create and have not thrown exception, then layer is supported
-            std::unique_ptr <MKLDNNNode>(MKLDNNNode::CreateNode(*i, eng, extensionManager, fake_w_cache));
-            res.supportedLayersMap.insert({ (*i)->name, GetName() });
-        } catch (InferenceEngine::details::InferenceEngineException&) {
+    MKLDNNWeightsSharing::Ptr fake_w_cache;
+    auto function = network.getFunction();
+    if (function != nullptr) {
+        std::unordered_set<std::string> originalOps;
+        for (auto&& node : function->get_ops()) {
+            if (!node->is_constant() && !node->is_parameter() && !node->is_output()) {
```
Comment (slyalin) [inline inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp:214]: Does it mean that all Constants (regardless of type and shape) and parameters (any kind of, even future extensions, like Parameters with default values or something like this; regardless of type and shape) are treated as always supported?

File: inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp [hunk 5]
```diff
@@ -203,18 +206,42 @@ void Engine::AddExtension(InferenceEngine::IExtensionPtr extension) {
...
+        auto clonedNetwork = cloneNetwork(network);
+        clonedNetwork = Transformation(clonedNetwork->getFunction(), clonedNetwork);
+        for (details::CNNNetworkIterator itLayer{clonedNetwork.get()}; itLayer != details::CNNNetworkIterator(); itLayer++) {
+            try {
+                mkldnn::engine eng(mkldnn::engine(mkldnn::engine::kind::cpu, 0));
+                // if we can create and have not thrown exception, then layer is supported
+                std::unique_ptr <MKLDNNNode>(MKLDNNNode::CreateNode(*itLayer, eng, extensionManager, fake_w_cache));
+                for (auto&& fusedLayerName : ngraph::getFusedNamesVector((*itLayer)->getNode())) {
+                    if (contains(originalOps, fusedLayerName)) {
+                        res.supportedLayersMap.emplace(fusedLayerName, GetName());
```
Comment (GlebKazantaev) [inline inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp:227]: Can you explain what the purpose of QueryNetworkResult?
Original ngraph operation before conversion can be decomposed to multiple operations and the same fusedLayerName may appear in multiple nodes. Is it expected?

File: inference-engine/src/hetero_plugin/hetero_executable_network.cpp [hunk 6]
```diff
@@ -312,14 +347,377 @@ HeteroExecutableNetwork::HeteroExecutableNetwork(const InferenceEngine::ICNNNetw
...
+        }),
+        std::end(orderedOps));
+    bool allEmpty = true;
+    for (auto&& node : orderedOps) {
+        auto& nodeInfo = node->get_rt_info();
+        auto itInfo = nodeInfo.find("affinity");
+        if (itInfo != nodeInfo.end()) {
+            queryNetworkResult.supportedLayersMap.emplace(
+                node->get_friendly_name(),
+                Parameter{itInfo->second}.as<std::string>());
```
Comment (GlebKazantaev) [inline inference-engine/src/hetero_plugin/hetero_executable_network.cpp:384]: We are not using ie Parameter for rt info attributes. May be it's better to use variant directly?

File: inference-engine/src/hetero_plugin/hetero_executable_network.cpp [hunk 7]
```diff
@@ -312,14 +347,377 @@ HeteroExecutableNetwork::HeteroExecutableNetwork(const InferenceEngine::ICNNNetw
...
+                result.emplace_back(std::move(input));
+            }
+        }
+        return result;
+    };
+
+    // Set parameters affinity
+    for (auto&& node : function->get_parameters()) {
+        if (!contains(queryNetworkResult.supportedLayersMap, node->get_friendly_name())) {
+            auto& outputNodeName = node->output(0).get_target_inputs().begin()->get_node()->get_friendly_name();
```
Comment (GlebKazantaev) [inline inference-engine/src/hetero_plugin/hetero_executable_network.cpp:419]: Why you check only one consumer? Parameter can have multiple consumers.

File: inference-engine/src/hetero_plugin/hetero_executable_network.cpp [hunk 8]
```diff
@@ -312,14 +347,377 @@ HeteroExecutableNetwork::HeteroExecutableNetwork(const InferenceEngine::ICNNNetw
...
+    // Break graph using insertion of result parameter split
+    NodeMap<ngraph::Node*> subgraphParameterToPrevResult;
+    std::vector<std::shared_ptr<ngraph::op::Result>> results;
+    for (auto&& input : subgraphInputs) {
+        if (!(input.get_node()->is_parameter())) {
+            auto output = input.get_source_output();
+            output.remove_target_input(input);
+            auto result = std::make_shared<ngraph::op::Result>(output);
+            result->set_placement(output.get_node()->get_placement());
+            ngraph::copy_runtime_info(output.get_node()->shared_from_this(), result);
```
Comment (GlebKazantaev) [inline inference-engine/src/hetero_plugin/hetero_executable_network.cpp:591]:
```suggestion
            ngraph::copy_runtime_info(output.get_node_shared_ptr(), result);
```

File: inference-engine/src/hetero_plugin/hetero_executable_network.cpp [hunk 9]
```diff
@@ -312,14 +347,377 @@ HeteroExecutableNetwork::HeteroExecutableNetwork(const InferenceEngine::ICNNNetw
...
+    auto subgraphIds = CollectSubgraphs();
+    // Break graph using insertion of result parameter split
+    NodeMap<ngraph::Node*> subgraphParameterToPrevResult;
+    std::vector<std::shared_ptr<ngraph::op::Result>> results;
+    for (auto&& input : subgraphInputs) {
+        if (!(input.get_node()->is_parameter())) {
+            auto output = input.get_source_output();
+            output.remove_target_input(input);
+            auto result = std::make_shared<ngraph::op::Result>(output);
+            result->set_placement(output.get_node()->get_placement());
```
Comment (GlebKazantaev) [inline inference-engine/src/hetero_plugin/hetero_executable_network.cpp:590]: Why you do this?

File: inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp (hunk 4 above)
Comment (apankratovantonp) [reply inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp:214]: No. This means that we do not take this nodes into account. It is supposed that they ofloaded to the same plugin as operation node with which they connected. I don't think that we have support of subgraph that conssists of single constant or parameter node in any plugin.

File: inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp [hunk 10]
```diff
@@ -49,6 +50,25 @@ Engine::~Engine() {
     ExecutorManager::getInstance()->clear("CPUCallbackExecutor");
 }
 
+static CNNNetworkImpl::Ptr Transformation(std::shared_ptr<::ngraph::Function> nGraphFunc,
```
[the remark this answers is not in the document]
Comment (apankratovantonp) [reply inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp:53]: This is the place for any transforamation not only conversion to legasy API

File: inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp (hunk 5 above)
Comment (apankratovantonp) [reply inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp:227]: Due to the legacy the most valuble QueryNetworkResult member is supportedLayersMap field. It contains maping of layer names to plugin names. QueryNetwork takes "Network" Operation as input. So QueryNetworkResult should contain just names from this operation set on output. No mater what intermediat transformation can be. The layer from "Network"  operation set is supported if it is used to form valid operation set for Plugin.

File: inference-engine/src/hetero_plugin/hetero_executable_network.cpp (hunk 6 above)
Comment (apankratovantonp) [reply inference-engine/src/hetero_plugin/hetero_executable_network.cpp:384]: Resently I was adwised to use ie::Parameter Here. @ilyachur @ilya-lavrenov what can you say?

File: inference-engine/src/hetero_plugin/hetero_executable_network.cpp (hunk 7 above)
Comment (apankratovantonp) [reply inference-engine/src/hetero_plugin/hetero_executable_network.cpp:419]: There is no selection cretaria for such case. So the first consumers affinity is good enogh here.

File: inference-engine/src/hetero_plugin/hetero_executable_network.cpp (hunk 9 above)
Comment (apankratovantonp) [reply inference-engine/src/hetero_plugin/hetero_executable_network.cpp:590]: Just copy past from `ngraph::insert_result_parameter_split` function.

Comment (apankratovantonp) [reply inference-engine/src/hetero_plugin/hetero_executable_network.cpp:590]: Should this code be deleted?

File: inference-engine/src/hetero_plugin/hetero_executable_network.cpp (hunk 8 above)
Comment (apankratovantonp) [reply inference-engine/src/hetero_plugin/hetero_executable_network.cpp]: Done

File: inference-engine/src/hetero_plugin/hetero_executable_network.cpp (hunk 6 above)
Comment (ilyachur) [reply inference-engine/src/hetero_plugin/hetero_executable_network.cpp:385]: @apankratovantonp This mechanism was refactored and all at the current moment transformations library doesn't contain any dependencies on IE. Moreover the rt propagation mechanism doesn't work with IE Parameter. @GlebKazantaev is right. Please change it to nGraph Variant.

File: inference-engine/src/hetero_plugin/hetero_executable_network.cpp (hunk 9 above)
Comment (ilyachur) [reply inference-engine/src/hetero_plugin/hetero_executable_network.cpp:596]: Yes the placement is a rudiment of nGraph code base and we don't use it in the Inference ENgine. please remove this code.

File: inference-engine/src/multi_device/multi_device.cpp [hunk 11]
```diff
@@ -525,24 +526,22 @@ void MultiDeviceInferencePlugin::QueryNetwork(const ICNNNetwork&
...
-    details::CNNNetworkIterator i(&network);
-    while (i != details::CNNNetworkIterator()) {
-        CNNLayer::Ptr layer = *i;
-        bool layerIsInQueryResultsForAllDevices = std::all_of(std::begin(queryResults), std::end(queryResults),
-            [&](const std::map<std::string, QueryNetworkResult>::value_type& qr) {
-                return qr.second.supportedLayersMap.end() != qr.second.supportedLayersMap.find(layer->name);});
-        if (layerIsInQueryResultsForAllDevices) {
-            queryResult.supportedLayersMap[layer->name] = GetName();
+        auto clonedNetwork = cloneNetwork(network);
+        auto deviceQr = GetCore()->QueryNetwork(*clonedNetwork, deviceName, metaDevice.config);
```
Comment (ilyachur) [inline inference-engine/src/multi_device/multi_device.cpp:534]: What happens if one plugin works with nGraph representation and another with legacy format?

File: inference-engine/src/hetero_plugin/hetero_executable_network.cpp (hunk 6 above)
Comment (apankratovantonp) [reply inference-engine/src/hetero_plugin/hetero_executable_network.cpp:385]: Done

File: inference-engine/src/multi_device/multi_device.cpp (hunk 11 above)
Comment (apankratovantonp) [reply inference-engine/src/multi_device/multi_device.cpp:534]: Added conversion to legacy implemenation if plugin without ngraph support found

File: inference-engine/src/hetero_plugin/hetero_plugin.cpp [hunk 12]
```diff
@@ -50,8 +56,40 @@ InferenceEngine::ExecutableNetworkInternal::Ptr Engine::LoadExeNetworkImpl(const
...
+        if (anyDeviceDoNotSupportNgraph) {
+            auto clonedNetwork = cloneNetwork(network);
+            auto function = clonedNetwork->getFunction();
+            ::ngraph::op::GenericIE::DisableReshape noReshape(function);
+            ::ngraph::pass::CommonOptimizations().run_on_function(function);
+            ::ngraph::pass::ConvertOpSet3ToOpSet2().run_on_function(function);
+            ::ngraph::pass::ConvertOpSet2ToOpSet1().run_on_function(function);
+            ::ngraph::pass::ConvertOpSet1ToLegacy().run_on_function(function);
+            return std::make_shared<HeteroExecutableNetwork>(
+                *InferenceEngine::details::convertFunctionToICNNNetwork(function, *clonedNetwork),
```
Comment (apankratovantonp) [inline inference-engine/src/hetero_plugin/hetero_plugin.cpp:85]: @ilyachur Is it correct way to convert to legacy representation?

Comment (ilyachur) [reply inference-engine/src/hetero_plugin/hetero_plugin.cpp:85]: Yes it is correct. First of all you need transform all operations to legacy and after that convert to old representation. It is correct!

**Comment by ilya-lavrenov** [comment]
@apankratovantonp please, resolve merge conflicts.

@ilyachur @slyalin @GlebKazantaev we are going to merge by EOW. Please, have a look.

**Comment by ilya-lavrenov** [comment]
@apankratovantonp please, resolve conflicts.

**Pull request merged**